### PR TITLE
#165020222 Report an article

### PR DESCRIPTION
**Pull request by niyobobo**
## What does this PR do?
Reporting an article that violates terms of the agreement, or any other kind of plagiarism.

## Description of Task to be completed?
- Add an endpoint to report articles.
- Add an endpoint to review reported articles.

## How should this be manually tested?
- Start the application with `npm start` and create an article to be able to test this functionality.
-  To report an article, make the following request.
```
 * URL      : /api/v1/report/articles/:slug
 * Method   : POST.
 * Header   : <auth token>
 * Params   : Slug of an article to be reported
 * Body     : { message: <Reason to report an article> }
```

- Please note that only admin will be able to access these endpoints

```
 * Get all reported articles
 *
 * URL      : /api/v1/report/articles
 * Method   : GET.
 * Header   :  <auth token>
```

## Any background context you want to provide?

- Only the authenticated user will be able to report an article.
- Only admin will have access to get reported articles.
- Only admin will take a decision on the reported article.
 To access administrator privileges, log in with email: `tester2@test.com` and password: `testuser`. 

## What are the relevant pivotal tracker stories?
[#165020222](https://www.pivotaltracker.com/n/projects/2321840/stories/165020222)

### CODE REVIEW

**Review by avpaul — DISMISSED** [review]
LGTM. 
But I have one suggestion: 
When reporting an article the URL is like this `URL: /api/v1/report/article/:slug` with *article in the singular form*, the same for *deleting a reported article.* But for getting reported articles we have *articles in plural form* `URL: /api/v1/report/articles`. I think we should keep using **articles in the plural form**.

**Comment by niyobobo** [comment]
@avpaul if you check well on the description of that endpoint, it's get all reported, not one. To get one it will be the same scenario of getting one normal article because you want to read that certain article to ckeck if the provided report message realy matches the content of the reported article, so I decided not to repeat myself because that API is already there.

**Comment by avpaul** [comment]
@niyobobo I'm not talking about getting one article or many! I'm talking about **how the API URL is written.** I was suggesting that the URL should be `URL: /api/v1/report/articles/:slug` for reporting an article or deleting one. As we are using the plural when referring to objects i.e `POST /api/v1/articles not  /api/v1/article` for creating an article.

File: controllers/reports.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    const { id: reporter } = req.user;
+    try {
+      const findArticle = await article.findOne({
+        where: { slug },
+        attributes: ['id', 'slug', 'title', 'body']
+      });
+      if (!findArticle) {
+        return res.status(404).send({
+          status: res.statusCode,
+          error: 'This article is not found',
```
Comment (aaronsekisambu) [inline controllers/reports.js]: @niyobobo I think this error message should be specific. We have a lot of "article not found" in our code. I think you could say this. "The article you're trying to report is not found"

File: controllers/reports.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        });
+      }
+
+      const findReport = await Report.findOne({ where: { reporter, articleId: findArticle.id } });
+      if (findReport) {
+        return res.status(400).send({
+          status: res.statusCode,
+          error: 'You have already reported this article',
```
Comment (aaronsekisambu) [inline controllers/reports.js:34]: @niyobobo could it be best to return the name of the reported in the error message: For example: "Aaron Sekisambu, you have already reported this article"

**Review by aaronsekisambu — CHANGES_REQUESTED** [review]
LGTMT fix those few comments and we are good to go.
I love the way you refactored the queries to avoid the for loop problems. Thanks for the initative.

File: controllers/reports.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
+import models from '../models';
+
+const { User, Report, article } = models;
+
+const reporting = {
```
Comment (shemaeric) [inline controllers/reports.js]: @niyobobo You can change this naming on reporting as it seems like continuous action.

**Review by shemaeric — CHANGES_REQUESTED** [review]
Looks good expect some changes, from the testing of the reporting functionalities it seems like a user is able even to report a draft article which I think is not good to report the article which is not even published. And also it would be better to reconsider the way you are handling the roles of the user. For the other developer working on the same project like me, I am struggling to Know who is the user with the role -1 or 0 or another number.

**Review by shemaeric — DISMISSED** [review]

**Review by shemaeric — CHANGES_REQUESTED** [review]
Looks good expect some changes, from the testing of the reporting functionalities it seems like a user is able even to report a draft article which I think is not good to report the article which is not even published. And also it would be better to reconsider the way you are handling the roles of the user. For the other developer working on the same project like me, I am struggling to Know who is the user with the role -1 or 0 or another number.

File: controllers/reports.js (hunk 3 above)
Comment (niyobobo) [reply controllers/reports.js]: @shemaeric changes applied

File: controllers/reports.js (hunk 1 above)
Comment (niyobobo) [reply controllers/reports.js]: > @niyobobo I'm not talking about getting one article or many! I'm talking about **how the API URL is written.** I was suggesting that the URL should be `URL: /api/v1/report/articles/:slug` for reporting an article or deleting one. As we are using the plural when referring to objects i.e `POST /api/v1/articles not /api/v1/article` for creating an article.

@avpaul  Changes applied

**Comment by niyobobo** [comment]
> @niyobobo I'm not talking about getting one article or many! I'm talking about **how the API URL is written.** I was suggesting that the URL should be `URL: /api/v1/report/articles/:slug` for reporting an article or deleting one. As we are using the plural when referring to objects i.e `POST /api/v1/articles not /api/v1/article` for creating an article.

@avpaul  Changes applied

File: controllers/reports.js (hunk 1 above)
Comment (niyobobo) [reply controllers/reports.js]: @aaronsekisambu changes applied

**Review by aaronsekisambu — DISMISSED** [review]

**Review by shemaeric — DISMISSED** [review]

**Review by avpaul — DISMISSED** [review]

**Review by fridolinho — DISMISSED** [review]
Good.

**Review by aaronsekisambu — APPROVED** [review]

**Review by avpaul — APPROVED** [review]

**Review by daud1 — APPROVED** [review]
👍 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏

**Pull request merged**